enterprise/server/util/ocicache/ocicache.go

```diff
@@ -38,8 +45,161 @@ const (
 	maxManifestSize = 10000000
 
 	cacheDigestFunction = repb.DigestFunction_SHA256
+
+	// Tag-to-digest mapper configuration
+	tagToDigestTTL        = 15 * time.Minute
+	tagToDigestMaxEntries = 1000
 )
 
+// tagToDigestEntry represents a cached tag-to-digest mapping with expiration.
+type tagToDigestEntry struct {
+	digest     string
+	expiration time.Time
+}
+
+// tagResolver caches tag-to-digest mappings to avoid repeated HEAD requests.
+type tagResolver struct {
+	lru   *lru.LRU[tagToDigestEntry]
+	mu    sync.RWMutex
+	clock clockwork.Clock
+}
+
+// NewTagResolver creates a new tag resolver that caches tag-to-digest mappings.
+func NewTagResolver(clock clockwork.Clock) interfaces.OCITagResolver {
+	lruCache, err := lru.NewLRU[tagToDigestEntry](&lru.Config[tagToDigestEntry]{
+		SizeFn:  func(_ tagToDigestEntry) int64 { return 1 },
+		MaxSize: tagToDigestMaxEntries,
+	})
+	if err != nil {
+		log.Errorf("Failed to initialize tag resolver: %s", err)
+		return &tagResolver{clock: clock}
+	}
+	return &tagResolver{
+		lru:   lruCache,
+		clock: clock,
+	}
+}
+
+// Register initializes and registers the OCI tag resolver with the environment.
+func Register(env *real_environment.RealEnv) error {
+	resolver := NewTagResolver(env.GetClock())
+	env.SetOCITagResolver(resolver)
+	return nil
+}
+
+// Get retrieves a non-expired digest for the given tag.
+// Returns empty string if not found or expired.
+// Automatically removes expired entries.
+func (r *tagResolver) Get(ctx context.Context, tag string) string {
+	if r == nil || r.lru == nil {
+		return ""
+	}
+
+	r.mu.RLock()
+	entry, ok := r.lru.Get(tag)
+	r.mu.RUnlock()
```

Comment:
missed this on the last round - `lru.Get` mutates the LRU (it updates the data structure so that `tag` will now be the MRU key) - so `RLock` here is unsafe.

I would also change `sync.RWMutex` to be a regular mutex to avoid reintroducing the mistake and make sure it doesn't exist elsewhere